src/crypto/tpm.rs

```diff
@@ -0,0 +1,200 @@
+        let (bytes_r, bytes_s) = signature.split_at(self.key_length);
+
+        let signature = Signature {
+            scheme: AsymSchemeUnion::ECDSA(self.hash_alg),
+            signature: SignatureData::EcdsaSignature {
+                r: bytes_r.to_vec(),
+                s: bytes_s.to_vec(),
+            },
+        };
+
+        let data = data.try_into().map_err(|_| {
+            CoseError::UnsupportedError("Invalid digest passed to verify".to_string())
+        })?;
+
+        let mut context = self.context.borrow_mut();
+
+        match context.verify_signature(self.key_handle, &data, signature) {
+            Ok(_) => Ok(true),
+            Err(tpm_error::Tss2Error(Tss2ResponseCode::FormatOne(FormatOneResponseCode(
+                TSS2_RC_SIGNATURE,
+            )))) => Ok(false),
+            Err(e) => Err(CoseError::TpmError(e)),
+        }
+    }
+}
+
+impl SigningPrivateKey for TpmKey {
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>, CoseError> {
+        let scheme = TPMT_SIG_SCHEME {
+            scheme: tpm_constants::tss::TPM2_ALG_NULL,
+            details: Default::default(),
+        };
+        let validation = TPMT_TK_HASHCHECK {
+            tag: tpm_constants::tss::TPM2_ST_HASHCHECK,
+            hierarchy: tpm_constants::tss::TPM2_RH_NULL,
+            digest: Default::default(),
+        };
+
+        let data = data
+            .try_into()
+            .map_err(|_| CoseError::UnsupportedError("Tried to sign invalid data".to_string()))?;
+
+        let signature = {
+            let mut context = self.context.borrow_mut();
+
+            context
+                .sign(
+                    self.key_handle,
+                    &data,
+                    scheme,
+                    validation.try_into().expect("Unable to convert validation"),
+                )
+                .map_err(CoseError::TpmError)?
+        };
+
+        let (bytes_r, bytes_s) = match signature.signature {
+            SignatureData::EcdsaSignature { r, s } => (r, s),
+            _ => {
+                return Err(CoseError::UnsupportedError(
+                    "Unsupported signature data returned".to_string(),
+                ))
+            }
+        };
+
+        assert!(bytes_r.len() <= self.key_length);
+        assert!(bytes_s.len() <= self.key_length);
+
+        let mut signature_bytes = vec![0u8; self.key_length * 2];
+
+        // This is big-endian encoding so padding might be added at the start if the factor is
+        // too short.
+        let offset_copy = self.key_length - bytes_r.len();
+        signature_bytes[offset_copy..offset_copy + bytes_r.len()].copy_from_slice(&bytes_r);
+
+        // This is big-endian encoding so padding might be added at the start if the factor is
+        // too short.
+        let offset_copy = self.key_length - bytes_s.len() + self.key_length;
+        signature_bytes[offset_copy..offset_copy + bytes_s.len()].copy_from_slice(&bytes_s);
+
+        Ok(signature_bytes)
```

Comment:
This code is also used in the OpenSSL PrivateKey and the KMS PrivateKey implementation. Should be refactored into a function.